### DISCUSSION TITLE
Quarkus support for AWS Lambda SnapStart

### DISCUSSION THREAD

**Discussion by giscus** [quarkusio/quarkus#29576] [Quarkus Blog/Website]
# Quarkus support for AWS Lambda SnapStart

Quarkus: Supersonic Subatomic Java

https://quarkus.io/blog/quarkus-support-for-aws-lambda-snapstart/

**Comment by drissamri** [comment]
Where can I find the test application used for the benchmarks?

**Reply by maxandersen** [reply]
I believe its https://github.com/quarkusio/quarkus-super-heroes that was used - @cescoffier can share details.

**Reply by cescoffier** [reply]
It's a simplified version of the super-heroes + a very simple hello world function.

**Comment by igor-dmitriev** [comment]
looks awesome

**Comment by saturna** [comment]
This might be a native-image killer - at least for AWS Lambdas specifically... I have ran this against my Quarkus lambda that uses DynamoDB, SSM and Okta SDK, and cold start went from 12 seconds down to 7 seconds, vs native image 4.5 seconds. But I wonder if I extract more stuff to be statically initialized maybe it will be even faster. TBD.

**Reply by maxandersen** [reply]
@saturna native-image has a bunch of benefits over snapstart (runs outside lambda, easier to test, more stable runtime performance as all code is compiled vs hotspot jit that needs to run a bit more) - but yes, due to how amazon lambda infrastructure runs and cost model native image is in some cases penalized despite using less memory and less cpu. 

It is an option for usecases that fits. Not a killer of native-image.

**Reply by cescoffier** [reply]
@saturna, there are a few features added to Quarkus:

- pre-load of classes (very user-unfriendly, but a huge step forward) - you would need to provide a file with the class used by your app, and Quarkus will preload all of them during the snapshot.
- priming of clients - using the preloading of classes or using a custom `@Startup` bean, you can pre-init your client, which also helps a lot. 

Unfortunately, these features are only in 2.15 (not yet released) and very under-documented (well... not documented at all).
